src/library/scala/collection/immutable/NVector.scala

```diff
@@ -0,0 +1,2010 @@
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.collection
+package immutable
+
+import scala.annotation.switch
+import scala.collection.mutable.{ArrayBuilder, Builder, ReusableBuilder}
+import scala.collection.generic.DefaultSerializable
+import scala.reflect.ClassTag
+import java.util.{Arrays, Spliterator}
+import java.util.Arrays.{copyOf, copyOfRange}
+import java.lang.Math.{abs, max => mmax, min => mmin}
+
+import NVectorStatics.{Arr1, Arr2, Arr3, Arr4, Arr5, Arr6}
+import scala.collection.Stepper.EfficientSplit
+
+
+object NVector extends StrictOptimizedSeqFactory[NVector] {
+  import NVectorStatics._
+
+  def empty[A]: NVector[A] = NVector0
+
+  def from[E](it: collection.IterableOnce[E]): NVector[E] =
+    it match {
+      case v: NVector[E] => v
+      case _ =>
+        val knownSize = it.knownSize
+        if (knownSize == 0) empty[E]
+        else if (knownSize > 0 && knownSize <= WIDTH) {
+          val a1 = new Array[Any](knownSize)
+          it.iterator.copyToArray(a1)
+          new NVector1[E](a1.asInstanceOf[Arr1])
+        } else {
+          (newBuilder ++= it).result()
+        }
+    }
+
+  def newBuilder[A]: ReusableBuilder[A, NVector[A]] = new NVectorBuilder[A]
+
+  /** Create a Vector with the same element at each index.
+    *
+    * Unlike `fill`, which takes a by-name argument for the value and can thereby
+    * compute different values for each index, this method guarantees that all
+    * elements are identical. This allows sparse allocation in O(log n) time and space.
+    */
+  def fillSparse[A](n: Int)(elem: A): NVector[A] = {
+    if(n <= 0) NVector0
+    else {
+      val b = new NVectorBuilder[A]
+      b.initSparse(n, elem)
+      b.result()
+    }
+  }
+}
+
+
+/** Vector is a general-purpose, immutable data structure.  It provides random access and updates
+  * in O(log n) time, as well as very fast append/prepend/tail/init (amortized O(1), worst case O(log n)).
+  * Because vectors strike a good balance between fast random selections and fast random functional updates,
+  * they are currently the default implementation of immutable indexed sequences.
+  *
+  * Vectors are implemented by radix-balanced finger trees of width 32. There is a separate subclass
+  * for each level (0 to 6, with 0 being the empty vector and 6 a tree with a maximum width of 64 at the
+  * top level).
+  *
+  * Tree balancing:
+  * - Only the first dimension of an array may have a size < WIDTH
+  * - In a `data` (central) array the first dimension may be up to WIDTH-2, in a prefix or suffix WIDTH-1
+  * - prefix1 and suffix1 are never empty
+  * - Balancing does not cross the main data array (i.e. prepending never touches the suffix and appending never touches
+  *   the prefix). The level is increased/decreased when the affected side plus main data is already full/empty
+  * - All arrays are left-aligned and truncated
```

Comment:
It would be good to encode some of these invariants in a `checkConsistent` method which we could call incorporate into a Scalacheck test suite that performed random operations and asserts the internal structure is consistent.

Reply:
There are validation methods in the JUnit test. I originally had them directly in `NVector` but factored them out to keep the code size down.